standard/patterns.md

```diff
@@ -157,13 +175,235 @@ var_pattern
 
 It is an error if the name `var` would bind to a type where a *var_pattern* is used.
 
+If *designation* is a *tuple_designation*, the pattern is equivalent to a *positional_pattern* (§positional-pattern-new-clause) of the form `(var` *designation*, ... `)` where the *designation*s are those found within the *tuple_designation*.  For example, the pattern `var (x, (y, z))` is equivalent to `(var x, (var y, var z))`.
+
+### §positional-pattern-new-clause Positional pattern
+
+A *positional_pattern* checks that the input value is not `null`, invokes an appropriate `Deconstruct` method ([§12.7](expressions.md#127-deconstruction)), and performs further pattern matching on the resulting values.  It also supports a tuple-like pattern syntax (without the type being provided) when the type of the input value is the same as the type containing `Deconstruct`, or if the type of the input value is a tuple type, or if the type of the input value is `object` or `System.ITuple` and the runtime type of the expression implements `System.ITuple`.
+
+```ANTLR
+positional_pattern
+    : type? '(' subpatterns? ')' property_subpattern? simple_designation?
+    ;
+subpatterns
+    : subpattern (',' subpatterns)?
+    ;
+subpattern
+    : pattern
+    | identifier ':' pattern
+    ;
+```
+
+Given a match of an input value to the pattern *type* `(` *subpatterns* `)`, a method is selected by searching in *type* for accessible declarations of `Deconstruct` and selecting one among them using the same rules as for the deconstruction declaration.
+It is an error if a *positional_pattern* omits the type, has a single *subpattern* without an *identifier*, has no *property_subpattern* and has no *simple_designation*. This disambiguates between a *constant_pattern* that is parenthesized and a *positional_pattern*.
+In order to extract the values to match against the patterns in the list,
+
+- If *type* is omitted and the input expression's type is a tuple type, then the number of subpatterns shall to be the same as the cardinality of the tuple. Each tuple element is matched against the corresponding *subpattern*, and the match succeeds if all of these succeed. If any *subpattern* has an *identifier*, then that shall name a tuple element at the corresponding position in the tuple type.
+- Otherwise, if a suitable `Deconstruct` exists as a member of *type*, it is a compile-time error if the type of the input value is not pattern-compatible with *type*. At runtime the input value is tested against *type*. If this fails, then the positional pattern match fails. If it succeeds, the input value is converted to this type and `Deconstruct` is invoked with fresh compiler-generated variables to receive the output parameters. Each value that was received is matched against the corresponding *subpattern*, and the match succeeds if all of these succeed. If any *subpattern* has an *identifier*, then that shall name a parameter at the corresponding position of `Deconstruct`.
```

Comment:
This is a low-level implementation description, there should be no talk of “compiler-generated variables” – that is an *implementation* detail.

We need to re-think how `Deconstruct` is woven into the description of pattern matching – as an approximation pattern matching may be performed on single (simple or compound) values, tuples of values (or types which implement `ITuple`?), and on compound values that offer to present themselves as “tuple-like“ (pick a suitable term) by providing `Deconstruct`.

If we can get this covered upfront the descriptions of the different patterns forms should be a lot simpler/clearer.